File: examples/C#/clonesrv1.cs

//
//  Clone server, Model One in C#
//

//  Author:     Yan Cui
//  Email:      theburningmonk@gmail.com

using System;
using System.Collections.Generic;
using System.Threading;
using ZMQ;

namespace ZMQGuide
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            using (var context = new Context(1))
            {
                using (var publisher = context.Socket(SocketType.PUB))
                {
                    publisher.Bind("tcp://*:5556");
                    Thread.Sleep(TimeSpan.FromMilliseconds(200));

                    var interrupted = false;
                    var sequence = 0L;
                    var random = new Random((int)DateTime.UtcNow.Ticks);
                    var dict = new Dictionary<string, KvMsg>();

                    Console.CancelKeyPress += (s, e) => { interrupted = true; };

                    while (!interrupted)
                    {
                        sequence++;
                        var kvmsg = new KvMsg(sequence);
                        kvmsg.Key = random.Next(1, 10000).ToString();
                        kvmsg.Body = random.Next(1, 1000000).ToString();

                        kvmsg.Send(publisher);
                        kvmsg.Store(dict);

                        Console.WriteLine("Published {0}", kvmsg);

                        Thread.Sleep(TimeSpan.FromMilliseconds(100));
                    }

                    Console.WriteLine(" Interrupted\n{0} messages out\n", sequence);
                    Console.ReadKey();
                }
            }
        }
    }
}

File: examples/Java/clonesrv1.java
package guide;

import java.nio.ByteBuffer;
import java.util.Random;
import java.util.concurrent.atomic.AtomicLong;

import org.zeromq.SocketType;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZContext;

/**
 * 
 * Clone server model 1
 * @author Danish Shrestha <dshrestha06@gmail.com>
 *
 */
public class clonesrv1
{
    private static AtomicLong sequence = new AtomicLong();

    public void run()
    {
        try (ZContext ctx = new ZContext()) {
            Socket publisher = ctx.createSocket(SocketType.PUB);
            publisher.bind("tcp://*:5556");

            try {
                Thread.sleep(200);
            }
            catch (InterruptedException e) {
                e.printStackTrace();
            }

            Random random = new Random();

            while (true) {
                long currentSequenceNumber = sequence.incrementAndGet();
                int key = random.nextInt(10000);
                int body = random.nextInt(1000000);

                ByteBuffer b = ByteBuffer.allocate(4);
                b.asIntBuffer().put(body);

                kvsimple kvMsg = new kvsimple(
                    key + "", currentSequenceNumber, b.array()
                );
                kvMsg.send(publisher);
                System.out.println("sending " + kvMsg);
            }
        }
    }

    public static void main(String[] args)
    {
        new clonesrv1().run();
    }
}
